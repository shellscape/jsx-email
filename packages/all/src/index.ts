export * from '@jsx-email/body';
export * from '@jsx-email/button';
export * from '@jsx-email/column';
export * from '@jsx-email/container';
export * from '@jsx-email/font';
export * from '@jsx-email/head';
export * from '@jsx-email/heading';
export * from '@jsx-email/hr';
export * from '@jsx-email/html';
export * from '@jsx-email/img';
export * from '@jsx-email/link';
export * from '@jsx-email/markdown';
export * from '@jsx-email/preview';
export * from '@jsx-email/render';
export * from '@jsx-email/row';
export * from '@jsx-email/section';
export * from '@jsx-email/tailwind';
export * from '@jsx-email/text';
