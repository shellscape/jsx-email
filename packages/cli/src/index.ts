#!/usr/bin/env node
import { parseArgs } from 'node:util';

import debugConfig from 'debug';
import importLocal from 'import-local';

import pkg from '../package.json';

import { command as build } from './commands/build';
import { command as create } from './commands/create';
import { command as help } from './commands/help';
import { command as preview } from './commands/preview';
import type { CommandFn } from './commands/types';

const commands: Record<string, CommandFn> = { build, create, help, preview };
const debug = debugConfig('@jsx-email/cli');
const { log } = console;

const run = async () => {
  const argv = parseArgs({ allowPositionals: true, args: process.argv.slice(2), strict: false });
  const [commandName] = argv.positionals;
  let command = commands[commandName];

  debug(`Command Name: \`${commandName}\``);

  if (argv.values.version) {
    log(`${pkg.name} v${pkg.version}\n`);
    return;
  }

  if (!command) command = help;

  const result = await command(argv.values, argv.positionals?.slice(1) || []);

  if (!result) {
    debug(`Command \`${commandName}\` returned \`false\``);
    help({}, []);
  }
};

if (importLocal(__filename)) {
  debug('Using local install of webpack-command');
} else {
  run();
}
