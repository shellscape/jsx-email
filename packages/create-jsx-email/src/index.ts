#!/usr/bin/env node
/* eslint-disable no-await-in-loop */
import { existsSync, readdirSync, rmSync } from 'fs';
import { mkdir, readFile, writeFile } from 'fs/promises';
import { basename, dirname, join, resolve, win32, posix } from 'path';

import chalk from 'chalk';
import { detect } from 'detect-package-manager';
import globby from 'globby';
import mustache from 'mustache';
import prompts from 'prompts';

import pkg from '../package.json';

interface CreateEmailArgs {
  jsx: boolean;
  name: string;
  outputPath: string;
}

const cancelled = () => {
  throw new Error(chalk`{red ✖} Operation cancelled`);
};
const defaultTargetDir = 'email-project';
const clearDirectory = (path: string) => {
  for (const file of readdirSync(path)) {
    // eslint-disable-next-line no-continue
    if (file === '.git') continue;
    rmSync(resolve(path, file), { force: true, recursive: true });
  }
};
const formatTargetDir = (targetPath?: string) => targetPath?.trim().replace(/\/+$/g, '');
const intro = chalk`
{underline create-jsx-email v${pkg.version}}

${pkg.description}
`;
const isEmpty = (path: string) => {
  const files = readdirSync(path);
  return files.length === 0 || (files.length === 1 && files[0] === '.git');
};
const { log } = console;
const normalizePath = (filename: string) => filename.split(win32.sep).join(posix.sep);
const typeDep = ',\n"@types/react": "^18.2.0",\n"typescript": "^5.2.2"';
const typeProps = `\nexport type TemplateProps = {
  email: string;
  name: string;
}`;

export const createEmail = async ({ jsx, name, outputPath }: CreateEmailArgs) => {
  const data = {
    name,
    propsType: jsx ? '' : ': TemplateProps',
    typeProps: jsx ? '' : typeProps
  };
  const templatePath = resolve(__dirname, '../generators/templates/email.mustache');
  const template = await readFile(templatePath, 'utf8');
  const contents = mustache.render(template, data);
  const fileName = basename(templatePath)
    .replace('_', '')
    .replace('.mustache', jsx ? '.jsx' : '.tsx');
  const outPath = join(outputPath, fileName);

  log('Creating a new template at', outPath);

  await writeFile(outPath, contents, 'utf8');

  return outPath;
};

const run = async () => {
  log(intro);

  const skip = process.argv.some((arg) => arg === '--yes');
  const argTargetDir = formatTargetDir(process.argv[2]);
  let targetPath = argTargetDir || defaultTargetDir;
  let result: prompts.Answers<'projectName' | 'projectType' | 'overwrite'>;
  const defaults: typeof result = {
    overwrite: true,
    projectName: defaultTargetDir,
    projectType: 'TypeScript'
  };

  try {
    result = skip
      ? defaults
      : await prompts(
          [
            {
              initial: defaultTargetDir,
              message: 'Project name:',
              name: 'projectName',
              onState: (state) => {
                targetPath = formatTargetDir(state.value) || defaultTargetDir;
              },
              type: argTargetDir ? null : 'text'
            },
            {
              choices: [
                { title: 'TypeScript', value: 'TypeScript' },
                { title: 'JavaScript', value: 'JavaScript' }
              ],
              message: 'TypeScript or JavaScript:',
              name: 'projectType',
              type: 'select'
            },
            {
              message: () =>
                targetPath === '.'
                  ? 'Current directory'
                  : `Target directory "${targetPath}" is not empty. Remove existing files and continue?`,
              name: 'overwrite',
              type: () => (!existsSync(targetPath) || isEmpty(targetPath) ? null : 'confirm')
            },
            {
              name: 'overwriteChecker',
              type: (_, { overwrite }: { overwrite?: boolean }) => {
                if (overwrite === false) cancelled();
                return null;
              }
            }
          ],
          { onCancel: cancelled }
        );
  } catch (cancelled: any) {
    log(cancelled.message);
    return;
  }

  const { overwrite, projectName, projectType } = result;
  const root = join(process.cwd(), targetPath);
  const generatorsPath = resolve(__dirname, '../generators');
  const jsx = projectType === 'JavaScript';
  const templates = await globby([normalizePath(join(generatorsPath, '/*.*'))]);
  const outputPath = join(root, 'templates');
  const templateData = { name: projectName, typeDep: jsx ? '' : typeDep };

  log(chalk`\n{blue Creating Project} at: {dim ${root}}`);

  if (overwrite && existsSync(root)) clearDirectory(root);

  await mkdir(outputPath, { recursive: true });

  for (const path of templates) {
    // eslint-disable-next-line no-continue
    if (path.endsWith('_tsconfig.json') && jsx) continue;
    const template = await readFile(path, 'utf8');
    const contents = mustache.render(template, templateData);
    const basePath = dirname(path);
    const fileName = basename(path).replace('_', '').replace('.mustache', '');
    const outPath = join(root, basePath.endsWith('templates') ? 'templates' : '', fileName);

    await writeFile(outPath, contents, 'utf8');
  }

  await createEmail({ jsx, name: projectName, outputPath });

  const packageManager = await detect();
  const install =
    packageManager === 'yarn'
      ? `  $ yarn\n  $ yarn dev`
      : `  $ ${packageManager} install\n  $ ${packageManager} run dev`;
  const footer = chalk`
{green ✓} {bold jsx-email} Project Created

Next, run:

  $ cd ${projectName}
${install}

{magenta Check out the docs!} {underline http://jsx.email/docs/quick-start}
`;

  log(footer);
};

run();
