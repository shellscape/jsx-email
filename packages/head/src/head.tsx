import * as React from 'react';

type HeadElement = React.ElementRef<'head'>;
type RootProps = React.ComponentPropsWithoutRef<'head'>;

export interface HeadProps extends RootProps {}

export const Head = React.forwardRef<HeadElement, Readonly<HeadProps>>(
  ({ children, ...props }, forwardedRef) => (
    <head {...props} ref={forwardedRef} data-id="@jsx-email/head">
      <meta httpEquiv="Content-Type" content="text/html; charset=UTF-8" />
      {children}
    </head>
  )
);

Head.displayName = 'Head';
