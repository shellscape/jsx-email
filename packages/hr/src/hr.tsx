import React from 'react';

type RootProps = React.ComponentPropsWithoutRef<'hr'>;

export interface HrProps extends RootProps {}

export const Hr: React.FC<Readonly<HrProps>> = ({ style, ...props }) => (
  <hr
    {...props}
    data-id="@jsx-email/hr"
    style={{
      border: 'none',
      borderTop: '1px solid #eaeaea',
      width: '100%',
      ...style
    }}
  />
);

Hr.displayName = 'Hr';
