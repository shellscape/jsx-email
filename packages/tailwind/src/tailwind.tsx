import * as React from 'react';
import { jsxToString } from '@jsx-email/render';
import { tailwindToCSS, type TailwindConfig } from 'tw-to-css';

import { cssToJsxStyle } from './css-to-jsx-style';

export interface TailwindProps {
  config?: TailwindConfig;
}

function processElement(
  element: React.ReactElement,
  headStyles: string[],
  config?: TailwindConfig
): React.ReactElement {
  const { twi } = tailwindToCSS({
    config
  });

  if (element.props.className) {
    const convertedStyles: string[] = [];
    const responsiveStyles: string[] = [];
    const classNames = element.props.className.split(' ');

    const customClassNames = classNames.filter((className: string) => {
      const tailwindClassNames = twi(className, { ignoreMediaQueries: true });
      if (tailwindClassNames) {
        convertedStyles.push(tailwindClassNames);
        return false;
      } else if (twi(className, { ignoreMediaQueries: false })) {
        responsiveStyles.push(className);
        return false;
      }
      return true;
    });

    const convertedResponsiveStyles = twi(responsiveStyles, {
      ignoreMediaQueries: false,
      merge: false
    });

    headStyles.push(convertedResponsiveStyles.replace(/^\n+/, '').replace(/\n+$/, ''));

    // eslint-disable-next-line no-param-reassign
    element = React.cloneElement(element, {
      ...element.props,
      // eslint-disable-next-line no-undefined
      className: customClassNames.length ? customClassNames.join(' ') : undefined,
      style: { ...element.props.style, ...cssToJsxStyle(convertedStyles.join(' ')) }
    });
  }

  if (element.props.children) {
    const children = React.Children.toArray(element.props.children);
    const processedChildren = children.map((child) => {
      if (React.isValidElement(child)) {
        return processElement(child, headStyles, config);
      }
      return child;
    });

    // eslint-disable-next-line no-param-reassign
    element = React.cloneElement(element, element.props, ...processedChildren);
  }

  return element;
}

function processHead(child: React.ReactElement, responsiveStyles: string[]): React.ReactElement {
  // FIXME: find a cleaner solution for child as any
  if (child.type === 'head' || (child as any).type.displayName === 'Head') {
    const styleElement = <style>{responsiveStyles}</style>;

    const headChildren = React.Children.toArray(child.props.children);
    headChildren.push(styleElement);

    // eslint-disable-next-line no-param-reassign
    child = React.cloneElement(child, child.props, ...headChildren);
  }
  if (child.props.children) {
    const children = React.Children.toArray(child.props.children);
    const processedChildren = children.map((processedChild) => {
      if (React.isValidElement(processedChild)) {
        return processHead(processedChild, responsiveStyles);
      }
      return processedChild;
    });

    // eslint-disable-next-line no-param-reassign
    child = React.cloneElement(child, child.props, ...processedChildren);
  }

  return child;
}

export const Tailwind = ({ children, config }: React.PropsWithChildren<TailwindProps>) => {
  const headStyles: string[] = [];

  const childrenWithInlineStyles = React.Children.map(children, (child) => {
    if (React.isValidElement(child)) {
      return processElement(child, headStyles, config);
    }
    return child;
  });

  if (!childrenWithInlineStyles) return <>{children}</>;

  const fullHTML = jsxToString(<>{childrenWithInlineStyles}</>);

  const hasResponsiveStyles = /@media[^{]+\{(?<content>[\s\S]+?)\}\s*\}/gm.test(
    headStyles.join(' ')
  );

  const hasHTMLAndHead = /<html[^>]*>(?=[\s\S]*<head[^>]*>)/gm.test(fullHTML);

  if (hasResponsiveStyles && !hasHTMLAndHead) {
    throw new Error(
      'Tailwind: To use responsive styles you must have a <html> and <head> element in your template.'
    );
  }

  const childrenWithInlineAndResponsiveStyles = React.Children.map(
    childrenWithInlineStyles,
    (child) => {
      if (React.isValidElement(child)) {
        return processHead(child, headStyles);
      }
      return child;
    }
  );

  return <>{childrenWithInlineAndResponsiveStyles}</>;
};
