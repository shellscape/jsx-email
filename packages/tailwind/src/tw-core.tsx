import { jsxToString } from '@jsx-email/render';
import chalk from 'chalk';
import { load } from 'cheerio';
// import { create } from 'twind';
import { defineConfig, extract, install } from '@twind/core';
// import { setup } from 'twind';
// import { virtualSheet, shim, getStyleTag, type Configuration } from 'twind/shim/server';
import presetAutoprefix from '@twind/preset-autoprefix';
import presetTailwind from '@twind/preset-tailwind';

const defaultConfig = defineConfig({
  presets: [presetAutoprefix(), presetTailwind()]
});

export interface TailwindProps {
  config?: typeof defaultConfig;
}

// @ts-ignore
const renderTwind = (html: string, config: TailwindProps['config']) => {
  // const sheet = virtualSheet();
  // const { tw } = create({ sheet, ...config });
  // setup({ sheet, ...config });
  // const shimmedHtml = shim(html, tw);
  const tw = install({ ...defaultConfig, ...config });
  // const shimmedHtml = shim(html, tw);
  const { html: shimmedHtml, css } = extract(html, tw);
  console.log(html);
  console.log(shimmedHtml);
  console.log(css);

  // const styleTag = getStyleTag(sheet);

  return { shimmedHtml, styleTag: `<style>${css}</style>` };
};

export const Tailwind = ({ children, config }: React.PropsWithChildren<TailwindProps>) => {
  const initialHtml = jsxToString(<>{children}</>);
  const { shimmedHtml, styleTag } = renderTwind(initialHtml, config);
  const $doc = load(shimmedHtml, { xml: { decodeEntities: false }, xmlMode: true });
  const $head = $doc('head');

  if ($head.length) {
    $head.append(styleTag);
  } else {
    const { warn } = console;
    const warning = chalk`
   Some email clients (like Gmail) do not support <style> elements within the document body.
   Additionally, media queries and responsive styles may not work properly
   See: {cyan https://www.caniemail.com/features/html-style}
   We recommend using the Tailwind component like so:
     <Tailwind><Head/><Body>...</Body></Tailwind>
`;
    warn(chalk`\n{yellow ⚠️  jsx-email:Tailwind → No <head> tag was found}${warning}`);

    $doc.root().prepend(styleTag);
  }

  const finalHtml = $doc.html()!.replace(/id="__twind"/g, `twind`);

  return <section dangerouslySetInnerHTML={{ __html: finalHtml }} />;
};
