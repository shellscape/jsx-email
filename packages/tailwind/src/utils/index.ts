export * from "./clean-css";
export * from "./get-media-query-css";
export * from "./make-css-map";
