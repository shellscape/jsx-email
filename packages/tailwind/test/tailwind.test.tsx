import { renderToStaticMarkup as render } from 'react-dom/server';
import { type TailwindConfig } from 'tw-to-css';
import { Hr } from '@jsx-email/hr';
import { Head } from '@jsx-email/head';
import { Html } from '@jsx-email/html';

import { Tailwind } from '../src';

describe('Tailwind component', () => {
  beforeEach(() => {
    vi.restoreAllMocks();
    vi.resetModules();
  });

  describe('Inline styles', () => {
    it('should render children with inline Tailwind styles', () => {
      const actualOutput = render(
        <Tailwind>
          <div className="bg-white text-sm" />
        </Tailwind>
      );

      expect(actualOutput).not.toBeNull();
    });
  });

  it('should be able to use background image', () => {
    const actualOutput = render(
      <Tailwind>
        <div className="bg-[url(https://example.com/image.png)]" />
      </Tailwind>
    );

    expect(actualOutput).toMatchSnapshot();
  });

  it('should override inline styles with Tailwind styles', () => {
    const actualOutput = render(
      <Tailwind>
        <div
          style={{ backgroundColor: 'red', fontSize: '12px' }}
          className="bg-black text-[16px]"
        />
      </Tailwind>
    );

    expect(actualOutput).toMatchSnapshot();
  });

  it('should override component styles with Tailwind styles', () => {
    const actualOutput = render(
      <Tailwind>
        <Hr className="w-12" />
      </Tailwind>
    );

    expect(actualOutput).toContain('width:3rem');
  });
});

describe('Responsive styles', () => {
  it('should add css to <head/>', () => {
    const actualOutput = render(
      <Tailwind>
        <html>
          <head />
          <body>
            <div className="bg-red-200 sm:bg-red-300 md:bg-red-400 lg:bg-red-500" />
          </body>
        </html>
      </Tailwind>
    );

    expect(actualOutput).toMatchSnapshot();
  });

  it('should throw an error when used without either <html/> or <head/> tags', () => {
    function noHtmlOrHead() {
      render(
        <Tailwind>
          <div className="bg-red-200 sm:bg-red-500" />
        </Tailwind>
      );
    }
    expect(noHtmlOrHead).toThrowErrorMatchingInlineSnapshot(
      `"Tailwind: To use responsive styles you must have a <html> and <head> element in your template."`
    );

    function noHtml() {
      render(
        <Tailwind>
          <head>
            <title>Test</title>
          </head>
          <div className="bg-red-200 sm:bg-red-500" />
        </Tailwind>
      );
    }
    expect(noHtml).toThrowErrorMatchingInlineSnapshot(
      `"Tailwind: To use responsive styles you must have a <html> and <head> element in your template."`
    );

    function noHead() {
      render(
        <Tailwind>
          <html>
            {/* <Head></Head> */}
            <div className="bg-red-200 sm:bg-red-500" />
          </html>
        </Tailwind>
      );
    }
    expect(noHead).toThrowErrorMatchingInlineSnapshot(
      `"Tailwind: To use responsive styles you must have a <html> and <head> element in your template."`
    );
  });

  it('should persist exsisting <head/> elements', () => {
    const actualOutput = render(
      <Tailwind>
        <html>
          <head>
            <style />
            <link />
          </head>
          <body>
            <div className="bg-red-200 sm:bg-red-500" />
          </body>
        </html>
      </Tailwind>
    );

    expect(actualOutput).toMatchSnapshot();
  });
});

describe('Custom theme config', () => {
  it('should be able to use custom colors', () => {
    const config: TailwindConfig = {
      theme: {
        extend: {
          colors: {
            custom: '#1fb6ff'
          }
        }
      }
    };

    const actualOutput = render(
      <Tailwind config={config}>
        <div className="text-custom bg-custom" />
      </Tailwind>
    );

    expect(actualOutput).toMatchSnapshot();
  });

  it('should be able to use custom colors', () => {
    const config: TailwindConfig = {
      theme: {
        extend: {
          fontFamily: {
            sans: ['Graphik', 'sans-serif'],
            serif: ['Merriweather', 'serif']
          }
        }
      }
    };

    const actualOutput = render(
      <Tailwind config={config}>
        <div className="font-sans" />
        <div className="font-serif" />
      </Tailwind>
    );

    expect(actualOutput).toMatchSnapshot();
  });

  it('should be able to use custom spacing', () => {
    const config: TailwindConfig = {
      theme: {
        extend: {
          spacing: {
            '8xl': '96rem'
          }
        }
      }
    };
    const actualOutput = render(
      <Tailwind config={config}>
        <div className="m-8xl"></div>
      </Tailwind>
    );
    expect(actualOutput).toMatchSnapshot();
  });

  it('should be able to use custom border radius', () => {
    const config: TailwindConfig = {
      theme: {
        extend: {
          borderRadius: {
            '4xl': '2rem'
          }
        }
      }
    };
    const actualOutput = render(
      <Tailwind config={config}>
        <div className="rounded-4xl" />
      </Tailwind>
    );
    expect(actualOutput).toMatchSnapshot();
  });

  it('should be able to use custom text alignment', () => {
    const config: TailwindConfig = {
      theme: {
        extend: {
          textAlign: {
            justify: 'justify'
          }
        }
      }
    };

    const actualOutput = render(
      <Tailwind config={config}>
        <div className="text-justify" />
      </Tailwind>
    );

    expect(actualOutput).toMatchSnapshot();
  });
});

describe('Custom plugins config', () => {
  it('should be able to use custom plugins', () => {
    const config: TailwindConfig = {
      plugins: [
        ({ addUtilities }: any) => {
          const newUtilities = {
            '.border-custom': {
              border: '2px solid'
            }
          };

          addUtilities(newUtilities);
        }
      ]
    };

    const actualOutput = render(
      <Tailwind config={config}>
        <div className="border-custom" />
      </Tailwind>
    );

    expect(actualOutput).toMatchSnapshot();
  });

  it('should be able to use custom plugins with responsive styles', () => {
    const config: TailwindConfig = {
      plugins: [
        ({ addUtilities }: any) => {
          const newUtilities = {
            '.border-custom': {
              border: '2px solid'
            }
          };

          addUtilities(newUtilities);
        }
      ]
    };

    const actualOutput = render(
      <Tailwind config={config}>
        <html>
          <head />
          <body>
            <div className="border-custom sm:border-custom" />
          </body>
        </html>
      </Tailwind>
    );

    expect(actualOutput).toMatchSnapshot();
  });
});

describe('<Tailwind> component', () => {
  it('should preserve mso styles', () => {
    const actualOutput = render(
      <Tailwind>
        <Html>
          <Head />
          <span
            dangerouslySetInnerHTML={{
              __html: `<!--[if mso]><i style="letter-spacing: 10px;mso-font-width:-100%;" hidden>&nbsp;</i><![endif]-->`
            }}
          />
          <div className="bg-white sm:bg-red-50 sm:text-sm md:text-lg custom-class" />
        </Html>
      </Tailwind>
    );

    expect(actualOutput).toMatchInlineSnapshot(
      '"<html data-id=\\"@jsx-email/html\\" lang=\\"en\\" dir=\\"ltr\\"><head data-id=\\"@jsx-email/head\\"><meta http-equiv=\\"Content-Type\\" content=\\"text/html; charset=UTF-8\\"/><style>@media(min-width:640px){.sm\\\\:text-sm{font-size:0.875rem;line-height:1.25rem}.sm\\\\:bg-red-50{background-color:rgb(254,242,242)}}@media(min-width:768px){.md\\\\:text-lg{font-size:1.125rem;line-height:1.75rem}}</style></head><span><!--[if mso]><i style=\\"letter-spacing: 10px;mso-font-width:-100%;\\" hidden>&nbsp;</i><![endif]--></span><div class=\\"custom-class\\" style=\\"background-color:rgb(255,255,255)\\"></div></html>"'
    );
  });

  it('should recognize custom resopnsive screen', () => {
    const config: TailwindConfig = {
      theme: {
        screens: {
          sm: { min: '640px' },
          // eslint-disable-next-line sort-keys
          md: { min: '768px' },
          // eslint-disable-next-line sort-keys
          lg: { min: '1024px' },
          xl: { min: '1280px' },
          // eslint-disable-next-line sort-keys
          '2xl': { min: '1536px' }
        }
      }
    };
    const actualOutput = render(
      <Tailwind config={config}>
        <Html>
          <Head />
          <div className="xl:bg-green-500">Test</div>
          <div className="2xl:bg-blue-500">Test</div>
        </Html>
      </Tailwind>
    );

    expect(actualOutput).toMatchInlineSnapshot(
      '"<html data-id=\\"@jsx-email/html\\" lang=\\"en\\" dir=\\"ltr\\"><head data-id=\\"@jsx-email/head\\"><meta http-equiv=\\"Content-Type\\" content=\\"text/html; charset=UTF-8\\"/><style>@media(min-width:1280px){.xl\\\\:bg-green-500{background-color:rgb(34,197,94)}}@media(min-width:1536px){.\\\\32xl\\\\:bg-blue-500{background-color:undefined}.\\\\32xl\\\\:bg-blue-500{background-color:rgb(59,130,246)}}</style></head><div>Test</div><div>Test</div></html>"'
    );
  });

  it('should work with calc() with + sign', () => {
    const actualOutput = render(
      <Tailwind>
        <div className="max-h-[calc(50px+3rem)] bg-red-100">
          <div className="h-[200px]">something tall</div>
        </div>
      </Tailwind>
    );

    expect(actualOutput).toMatchInlineSnapshot(
      '"<div style=\\"max-height:calc(50px + 3rem);background-color:rgb(254,226,226)\\"><div style=\\"height:200px\\">something tall</div></div>"'
    );
  });
});
