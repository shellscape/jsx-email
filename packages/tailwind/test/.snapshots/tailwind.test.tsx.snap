// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Custom plugins config > should be able to use custom plugins 1`] = `"<div style=\\"border:2px solid\\"></div>"`;

exports[`Custom plugins config > should be able to use custom plugins with responsive styles 1`] = `"<html><head><style>@media(min-width:640px){.sm\\\\:border-custom{border:2px solid}}</style></head><body><div style=\\"border:2px solid\\"></div></body></html>"`;

exports[`Custom theme config > should be able to use custom border radius 1`] = `"<div style=\\"border-radius:2rem\\"></div>"`;

exports[`Custom theme config > should be able to use custom colors 1`] = `"<div style=\\"color:rgb(31,182,255);background-color:rgb(31,182,255)\\"></div>"`;

exports[`Custom theme config > should be able to use custom colors 2`] = `"<div style=\\"font-family:Graphik, sans-serif\\"></div><div style=\\"font-family:Merriweather, serif\\"></div>"`;

exports[`Custom theme config > should be able to use custom spacing 1`] = `"<div style=\\"margin:96rem\\"></div>"`;

exports[`Custom theme config > should be able to use custom text alignment 1`] = `"<div style=\\"text-align:justify\\"></div>"`;

exports[`Responsive styles > should add css to <head/> 1`] = `"<html><head><style>@media(min-width:640px){.sm\\\\:bg-red-300{background-color:rgb(252,165,165)}}@media(min-width:768px){.md\\\\:bg-red-400{background-color:rgb(248,113,113)}}@media(min-width:1024px){.lg\\\\:bg-red-500{background-color:rgb(239,68,68)}}</style></head><body><div style=\\"background-color:rgb(254,202,202)\\"></div></body></html>"`;

exports[`Responsive styles > should persist exsisting <head/> elements 1`] = `"<html><head><style></style><link/><style>@media(min-width:640px){.sm\\\\:bg-red-500{background-color:rgb(239,68,68)}}</style></head><body><div style=\\"background-color:rgb(254,202,202)\\"></div></body></html>"`;

exports[`Tailwind component > should be able to use background image 1`] = `"<div style=\\"background-image:url(https://example.com/image.png)\\"></div>"`;

exports[`Tailwind component > should override inline styles with Tailwind styles 1`] = `"<div style=\\"background-color:rgb(0,0,0);font-size:16px\\"></div>"`;
